File: packages/python/tests/behavior_runner.py
"""
Runner for the normative behavioral suite, spec/behavior-vectors.json
(SPECIFICATION.md [N-47], [N-49]).

The scenarios are data. This file is the only thing that is language-specific,
which is what stops the ten ports from drifting apart the way ten hand-written
suites did.
"""

from __future__ import annotations

import json
from dataclasses import asdict, dataclass, field
from pathlib import Path
from typing import Any, Optional, Union

from nebula_token import (
    IssueResult,
    MemoryRefreshTokenStore,
    NebulaEngine,
    RefreshError,
    RefreshOk,
    RevokeError,
    TokenRecord,
    TokenStatus,
)


def _repo_root() -> Path:
    """Walk up from this file to the repository root ([N-49] artifacts live in
    <root>/spec). Never hardcoded, never vendored into the package."""
    for parent in Path(__file__).resolve().parents:
        if (parent / "spec" / "behavior-vectors.json").is_file():
            return parent
    raise RuntimeError("spec/behavior-vectors.json not found in any parent of this file")


SPEC_DIR = _repo_root() / "spec"

Vectors = dict[str, Any]
Scenario = dict[str, Any]
Step = dict[str, Any]


def load_behavior_vectors() -> Vectors:
    data: Vectors = json.loads((SPEC_DIR / "behavior-vectors.json").read_text(encoding="utf-8"))
    return data


#: Conditions this runtime satisfies. CPython's `str` is a sequence of code
#: points and can hold an unpaired surrogate — `json.loads(r'"\ud800"')` returns
#: one — so the invalid-Unicode scenario applies here.
SATISFIED_CONDITIONS = frozenset({"runtime-admits-invalid-unicode-strings"})

#: 32 zero bytes, canonically encoded: well-formed, and never the real secret.
FORGED_VERIFIER = "A" * 43
FORGED_SELECTOR = "A" * 22
LONE_SURROGATE = "\ud800"


class ControllableStore:
    """Wraps the reference store so a scenario can force one compare-and-set to lose."""

    def __init__(self) -> None:
        self.inner = MemoryRefreshTokenStore()
        self._fail_next: set[str] = set()

    def fail_next_cas(self, method: str) -> None:
        self._fail_next.add(method)

    def _consume(self, method: str) -> bool:
        if method in self._fail_next:
            self._fail_next.discard(method)
            return True
        return False

    def find_by_selector(self, selector: str) -> Optional[TokenRecord]:
        return self.inner.find_by_selector(selector)

    def insert(self, record: TokenRecord) -> None:
        self.inner.insert(record)

    def mark_rotated(
        self,
        selector: str,
        from_status: TokenStatus,
        rotated_at: int,
        replaced_by_selector: str,
    ) -> bool:
        if self._consume("markRotated"):
            return False
        return self.inner.mark_rotated(selector, from_status, rotated_at, replaced_by_selector)

    def revoke_if_active(self, selector: str) -> bool:
        if self._consume("revokeIfActive"):
            return False
        return self.inner.revoke_if_active(selector)

    def revoke_family(self, family_id: str) -> int:
        return self.inner.revoke_family(family_id)

    def revoke_user(self, user_id: str) -> int:
        return self.inner.revoke_user(user_id)


@dataclass(frozen=True)
class Binding:
    token: str
    family_id: str
    expires_at: int


@dataclass(frozen=True)
class SkippedScenario:
    id: str
    condition: str


@dataclass
class RunOutcome:
    executed: list[str] = field(default_factory=list)
    skipped: list[SkippedScenario] = field(default_factory=list)


class ScenarioFailure(AssertionError):
    """A divergence from a published scenario."""


def _fail(scenario: Scenario, index: int, message: str) -> "ScenarioFailure":
    requirements = ", ".join(scenario.get("requirements", []))
    return ScenarioFailure(f"[{scenario['id']}] step {index} ({requirements}): {message}")


def is_applicable(scenario: Scenario) -> bool:
    condition = scenario.get("condition")
    return condition is None or condition in SATISFIED_CONDITIONS


def run_behavior_vectors(vectors: Vectors) -> RunOutcome:
    """Execute every applicable scenario. Raises on the first divergence."""
    outcome = RunOutcome()
    for scenario in vectors["scenarios"]:
        if not is_applicable(scenario):
            outcome.skipped.append(SkippedScenario(scenario["id"], scenario["condition"]))
            continue
        run_scenario(vectors, scenario)
        outcome.executed.append(scenario["id"])
    return outcome


class _Clock:
    def __init__(self, now: int) -> None:
        self.now = now

    def __call__(self) -> int:
        return self.now


def run_scenario(vectors: Vectors, scenario: Scenario) -> None:
    cfg: dict[str, Any] = {**vectors["defaults"], **scenario.get("config", {})}
    store = ControllableStore()
    bindings: dict[str, Binding] = {}
    issued_secrets: list[str] = []
    device_ids: set[str] = set()
    clock = _Clock(cfg["now"])

    def build(kids: list[str], active_kid: str) -> NebulaEngine:
        return NebulaEngine(
            peppers={kid: vectors["peppers"][kid] for kid in kids},
            active_kid=active_kid,
            store=store,
            absolute_ttl_seconds=cfg["absoluteTtlSeconds"],
            idle_ttl_seconds=cfg["idleTtlSeconds"],
            reuse_grace_seconds=cfg["reuseGraceSeconds"],
            clock=clock,
        )

    engine = build(cfg["peppers"], cfg["activeKid"])

    def resolve_token(step: Step, index: int) -> str:
        ref: dict[str, Any] = step.get("token") or {}
        if "literal" in ref:
            literal: str = ref["literal"]
            return literal
        if "ref" not in ref:
            raise _fail(scenario, index, "step has no token reference")
        bound = bindings.get(ref["ref"])
        if bound is None:
            raise _fail(scenario, index, f"unknown binding {ref['ref']!r}")
        forge = ref.get("forge")
        if forge is None:
            return bound.token
        parts = bound.token.split(".")
        if forge == "verifier":
            parts[3] = FORGED_VERIFIER
        elif forge == "unknownKid":
            parts[1] = "zz"
        elif forge == "unknownSelector":
            parts[2] = FORGED_SELECTOR
        else:
            raise _fail(scenario, index, f"unknown forge {forge!r}")
        return ".".join(parts)

    def device_of(step: Step) -> Optional[str]:
        # Absence and the empty string are different values ([N-25]), so this
        # must not collapse a missing key into "".
        if step.get("deviceIdKind") == "lone-surrogate":
            return LONE_SURROGATE
        device_id: Optional[str] = step.get("deviceId")
        return device_id

    def check_success(
        result: Union[IssueResult, RefreshOk], expect: Optional[dict[str, Any]], index: int
    ) -> None:
        if not expect:
            return
        if "generation" in expect and result.generation != expect["generation"]:
            raise _fail(
                scenario, index, f"expected generation {expect['generation']}, got {result.generation}"
            )
        if "kid" in expect:
            kid = result.token.split(".")[1]
            if kid != expect["kid"]:
                raise _fail(scenario, index, f"expected kid {expect['kid']}, got {kid}")
        if "sameFamilyAs" in expect:
            other = bindings[expect["sameFamilyAs"]]
            if result.family_id != other.family_id:
                raise _fail(scenario, index, "family_id changed across rotation")
        if "sameExpiresAtAs" in expect:
            other = bindings[expect["sameExpiresAtAs"]]
            if result.expires_at != other.expires_at:
                raise _fail(
                    scenario,
                    index,
                    f"absolute deadline moved: {other.expires_at} -> {result.expires_at}",
                )
        if expect.get("idleEqualsExpires") and result.idle_expires_at != result.expires_at:
            raise _fail(
                scenario,
                index,
                f"idle_expires_at {result.idle_expires_at} should be clamped to {result.expires_at}",
            )

    def check_attribution(
        result: Union[RefreshError, RevokeError], expect: Optional[dict[str, Any]], index: int
    ) -> None:
        """[N-39] attribution, tri-state. `true` demands the field, `false`
        demands its absence — the exclusion list (MALFORMED, UNKNOWN_KID,
        NOT_FOUND) is a requirement too, and a truthy-only check could never
        observe it. Absent means the scenario does not assert it. Here a field
        is "absent" when it is None, which is how the engine signals that no
        record was resolved."""
        if expect is None:
            return
        want_user = expect.get("hasUserId")
        if want_user is not None and (result.user_id is not None) != want_user:
            raise _fail(
                scenario, index, f"expected user_id {'present' if want_user else 'absent'} ([N-39])"
            )
        want_family = expect.get("hasFamilyId")
        if want_family is not None and (result.family_id is not None) != want_family:
            raise _fail(
                scenario,
                index,
                f"expected family_id {'present' if want_family else 'absent'} ([N-39])",
            )

    def bind(step: Step, result: Union[IssueResult, RefreshOk]) -> None:
        name = step.get("bind")
        if name is not None:
            bindings[name] = Binding(result.token, result.family_id, result.expires_at)
        issued_secrets.append(result.token.split(".")[3])

    for index, step in enumerate(scenario["steps"]):
        expect: Optional[dict[str, Any]] = step.get("expect")
        op = step["op"]

        if op == "issue":
            device_id = device_of(step)
            issued = engine.issue(step["userId"], device_id)
            if expect is not None and expect.get("ok") is False:
                raise _fail(scenario, index, "expected issue to fail")
            check_success(issued, expect, index)
            bind(step, issued)
            if device_id:
                device_ids.add(device_id)

        elif op == "refresh":
            result = engine.refresh(resolve_token(step, index), device_of(step))
            wants_success = expect is None or (
                expect.get("ok") is True or ("ok" not in expect and "error" not in expect)
            )
            if wants_success:
                if not result.ok:
                    raise _fail(scenario, index, f"expected success, got {result.error}")
                check_success(result, expect, index)
                bind(step, result)
            else:
                assert expect is not None
                if result.ok:
                    raise _fail(scenario, index, f"expected {expect.get('error')}, got success")
                if result.error != expect.get("error"):
                    raise _fail(
                        scenario, index, f"expected {expect.get('error')}, got {result.error}"
                    )
                check_attribution(result, expect, index)

        elif op == "revokeToken":
            revoked = engine.revoke_token(resolve_token(step, index))
            if expect is not None and expect.get("ok") is False:
                if revoked.ok:
                    raise _fail(scenario, index, f"expected {expect.get('error')}, got success")
                if revoked.error != expect.get("error"):
                    raise _fail(
                        scenario, index, f"expected {expect.get('error')}, got {revoked.error}"
                    )
                # [N-39] governs every failure result, revoke_token's included.
                check_attribution(revoked, expect, index)
            else:
                if not revoked.ok:
                    raise _fail(scenario, index, f"expected success, got {revoked.error}")
                if expect is not None and "revoked" in expect and revoked.revoked != expect["revoked"]:
                    raise _fail(
                        scenario, index, f"expected {expect['revoked']} revoked, got {revoked.revoked}"
                    )

        elif op == "revokeFamilyOf":
            count = engine.revoke_family(bindings[step["of"]].family_id)
            if expect is not None and "revoked" in expect and count != expect["revoked"]:
                raise _fail(scenario, index, f"expected {expect['revoked']} revoked, got {count}")

        elif op == "revokeUser":
            count = engine.revoke_all_for_user(step["userId"])
            if expect is not None and "revoked" in expect and count != expect["revoked"]:
                raise _fail(scenario, index, f"expected {expect['revoked']} revoked, got {count}")

        elif op == "advance":
            clock.now += step["seconds"]

        elif op == "reconfigure":
            engine = build(step["peppers"], step["activeKid"])

        elif op == "failNextCas":
            store.fail_next_cas(step["method"])

        elif op == "expectStatusCounts":
            actual: dict[str, int] = {"active": 0, "rotated": 0, "revoked": 0}
            for row in store.inner.all():
                actual[row.status] += 1
            for status, want in step["counts"].items():
                if actual[status] != want:
                    raise _fail(
                        scenario, index, f"expected {want} {status}, got {actual[status]} ({actual})"
                    )

        elif op == "expectNoRawSecrets":
            # asdict, not repr: the record's repr redacts hashes ([N-14]), and a
            # redacted dump would make this assertion vacuous.
            dump = json.dumps([asdict(row) for row in store.inner.all()])
            for secret in issued_secrets:
                if secret in dump:
                    raise _fail(scenario, index, "a raw verifier reached the store ([N-14])")
            for device_id_value in device_ids:
                if device_id_value in dump:
                    raise _fail(scenario, index, "a raw device identifier reached the store ([N-14])")

        else:
            raise _fail(scenario, index, f"unknown op {op!r}")

File: packages/java/src/test/java/dev/nebulatoken/BehaviorRunner.java
package dev.nebulatoken;

import com.fasterxml.jackson.databind.JsonNode;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.LinkedHashMap;
import java.util.LinkedHashSet;
import java.util.List;
import java.util.Map;
import java.util.Set;
import java.util.concurrent.ConcurrentHashMap;
import java.util.concurrent.atomic.AtomicLong;

/**
 * Runner for the normative behavioral suite, spec/behavior-vectors.json
 * ([N-47], [N-49]).
 *
 * <p>The scenarios are data. This file is the only language-specific part, which
 * is what stops the ten ports from drifting apart the way ten hand-written
 * suites did. Nothing here re-derives a case; if a behaviour is not in the
 * vectors it belongs in {@link EngineTest}, not here.
 */
final class BehaviorRunner {

    /**
     * Conditions this runtime satisfies. Java strings are UTF-16 code-unit
     * sequences and can hold an unpaired surrogate, so the invalid-Unicode
     * scenario applies here and is executed rather than skipped.
     */
    static final Set<String> SATISFIED_CONDITIONS = Set.of("runtime-admits-invalid-unicode-strings");

    /** 32 zero bytes, canonically encoded: well-formed, and never the real secret. */
    private static final String FORGED_VERIFIER = "A".repeat(Nebula.VERIFIER_CHARS);
    private static final String FORGED_SELECTOR = "A".repeat(Nebula.SELECTOR_CHARS);
    private static final String LONE_SURROGATE = "\uD800";

    private BehaviorRunner() {}

    record Skipped(String id, String condition) {}

    record RunOutcome(List<String> executed, List<Skipped> skipped) {}

    static JsonNode load() {
        return SpecVectors.load("behavior-vectors.json");
    }

    /** Execute every applicable scenario. Throws on the first divergence. */
    static RunOutcome run(JsonNode vectors) {
        List<String> executed = new ArrayList<>();
        List<Skipped> skipped = new ArrayList<>();

        for (JsonNode scenario : vectors.get("scenarios")) {
            String condition = scenario.has("condition") ? scenario.get("condition").asText() : null;
            if (condition != null && !SATISFIED_CONDITIONS.contains(condition)) {
                skipped.add(new Skipped(scenario.get("id").asText(), condition));
                continue;
            }
            runScenario(vectors, scenario);
            executed.add(scenario.get("id").asText());
        }
        return new RunOutcome(executed, skipped);
    }

    /**
     * Cognitive complexity is high and stays high (java:S3776).
     *
     * <p>This method is an interpreter for the step vocabulary published in
     * spec/behavior-vectors.json. Every {@code case} below is one {@code op} in
     * that file, in the order the file lists them, so the vocabulary and the
     * code that executes it can be diffed against each other by eye -- which is
     * the property that makes a conformance runner auditable at all.
     *
     * <p>Extracting one method per op was considered and rejected. The ops share
     * one mutable scenario state -- {@code engine}, {@code store},
     * {@code bindings}, {@code issuedSecrets}, {@code presentedDeviceIds},
     * {@code now} -- and {@code reconfigure} <em>reassigns</em> {@code engine},
     * which an extracted static method cannot do without a holder object. The
     * result would be a dozen methods each taking a context parameter, the same
     * total branching, and a dispatch table that no longer reads like the vector
     * format. That trades an honest measurement for a worse artefact.
     */
    @SuppressWarnings("java:S3776")
    static void runScenario(JsonNode vectors, JsonNode scenario) {
        JsonNode defaults = vectors.get("defaults");
        JsonNode override = scenario.get("config");

        AtomicLong now = new AtomicLong(setting(defaults, override, "now").asLong());
        long absoluteTtl = setting(defaults, override, "absoluteTtlSeconds").asLong();
        long idleTtl = setting(defaults, override, "idleTtlSeconds").asLong();
        long grace = setting(defaults, override, "reuseGraceSeconds").asLong();

        ControllableStore store = new ControllableStore();
        Map<String, Binding> bindings = new LinkedHashMap<>();
        List<String> issuedSecrets = new ArrayList<>();
        Set<String> presentedDeviceIds = new LinkedHashSet<>();

        NebulaEngine engine = build(vectors, store, now, absoluteTtl, idleTtl, grace,
                strings(setting(defaults, override, "peppers")),
                setting(defaults, override, "activeKid").asText());

        int index = -1;
        for (JsonNode step : scenario.get("steps")) {
            index++;
            JsonNode expect = step.get("expect");

            switch (step.get("op").asText()) {
                case "issue" -> {
                    String deviceId = deviceOf(step);
                    IssueResult res = engine.issue(step.get("userId").asText(), deviceId);
                    if (expect != null && expect.has("ok") && !expect.get("ok").asBoolean()) {
                        throw fail(scenario, index, "expected issue to fail");
                    }
                    checkSuccess(scenario, index, bindings, expect,
                            res.token(), res.generation(), res.familyId(),
                            res.expiresAt(), res.idleExpiresAt());
                    bind(bindings, step, res.token(), res.familyId(), res.expiresAt());
                    issuedSecrets.add(verifierPartOf(res.token()));
                    if (deviceId != null && !deviceId.isEmpty()) presentedDeviceIds.add(deviceId);
                }

                case "refresh" -> {
                    String deviceId = deviceOf(step);
                    RefreshResult res = engine.refresh(
                            resolveToken(scenario, index, bindings, step.get("token")), deviceId);
                    if (deviceId != null && !deviceId.isEmpty()) presentedDeviceIds.add(deviceId);

                    if (expectsSuccess(expect)) {
                        if (res instanceof RefreshResult.Failure f) {
                            throw fail(scenario, index, "expected success, got " + f.error());
                        }
                        RefreshResult.Success ok = (RefreshResult.Success) res;
                        checkSuccess(scenario, index, bindings, expect,
                                ok.token(), ok.generation(), ok.familyId(),
                                ok.expiresAt(), ok.idleExpiresAt());
                        bind(bindings, step, ok.token(), ok.familyId(), ok.expiresAt());
                        issuedSecrets.add(verifierPartOf(ok.token()));
                    } else {
                        String want = expect.get("error").asText();
                        if (!(res instanceof RefreshResult.Failure f)) {
                            throw fail(scenario, index, "expected " + want + ", got success");
                        }
                        if (!f.error().name().equals(want)) {
                            throw fail(scenario, index, "expected " + want + ", got " + f.error());
                        }
                        checkAttribution(scenario, index, expect, f.userId(), f.familyId());
                    }
                }

                case "revokeToken" -> {
                    RevokeResult res = engine.revokeToken(
                            resolveToken(scenario, index, bindings, step.get("token")));
                    if (expect != null && expect.has("ok") && !expect.get("ok").asBoolean()) {
                        String want = expect.get("error").asText();
                        if (!(res instanceof RevokeResult.Failure f)) {
                            throw fail(scenario, index, "expected " + want + ", got success");
                        }
                        if (!f.error().name().equals(want)) {
                            throw fail(scenario, index, "expected " + want + ", got " + f.error());
                        }
                        // [N-39] governs every failure result, revokeToken's included.
                        checkAttribution(scenario, index, expect, f.userId(), f.familyId());
                    } else {
                        if (res instanceof RevokeResult.Failure f) {
                            throw fail(scenario, index, "expected success, got " + f.error());
                        }
                        checkRevoked(scenario, index, expect, ((RevokeResult.Success) res).revoked());
                    }
                }

                case "revokeFamilyOf" -> {
                    Binding of = bindings.get(step.get("of").asText());
                    if (of == null) throw fail(scenario, index, "unknown binding " + step.get("of").asText());
                    checkRevoked(scenario, index, expect, engine.revokeFamily(of.familyId()));
                }

                case "revokeUser" ->
                        checkRevoked(scenario, index, expect,
                                engine.revokeAllForUser(step.get("userId").asText()));

                case "advance" -> now.addAndGet(step.get("seconds").asLong());

                // A new engine over the SAME store: pepper rotation is a config
                // change, not a data migration.
                case "reconfigure" -> engine = build(vectors, store, now, absoluteTtl, idleTtl, grace,
                        strings(step.get("peppers")), step.get("activeKid").asText());

                case "failNextCas" -> store.failNextCas(step.get("method").asText());

                case "expectStatusCounts" -> {
                    Map<String, Integer> actual = new LinkedHashMap<>();
                    for (TokenStatus status : TokenStatus.values()) {
                        actual.put(status.name().toLowerCase(java.util.Locale.ROOT), 0);
                    }
                    for (TokenRecord row : store.inner.all()) {
                        String key = row.status().name().toLowerCase(java.util.Locale.ROOT);
                        actual.merge(key, 1, Integer::sum);
                    }
                    JsonNode counts = step.get("counts");
                    var names = counts.fieldNames();
                    while (names.hasNext()) {
                        String status = names.next();
                        int want = counts.get(status).asInt();
                        if (!Integer.valueOf(want).equals(actual.get(status))) {
                            throw fail(scenario, index, "expected " + want + " " + status
                                    + ", got " + actual.get(status) + " " + actual);
                        }
                    }
                }

                case "expectNoRawSecrets" -> {
                    String dump = store.inner.all().toString();
                    for (String secret : issuedSecrets) {
                        if (dump.contains(secret)) {
                            throw fail(scenario, index, "a raw verifier reached the store ([N-14])");
                        }
                    }
                    for (String deviceId : presentedDeviceIds) {
                        if (dump.contains(deviceId)) {
                            throw fail(scenario, index, "a raw device identifier reached the store ([N-14])");
                        }
                    }
                }

                default -> throw fail(scenario, index, "unknown op \"" + step.get("op").asText() + "\"");
            }
        }
    }

    // --- step helpers -------------------------------------------------------

    private record Binding(String token, String familyId, long expiresAt) {}

    /** A step with no {@code expect}, or one naming neither ok nor error, expects success. */
    private static boolean expectsSuccess(JsonNode expect) {
        if (expect == null) return true;
        if (expect.has("ok")) return expect.get("ok").asBoolean();
        return !expect.has("error");
    }

    /**
     * Nine parameters (java:S107), and each one is a field of the vector step or
     * of the result being checked against it. Bundling them into a carrier type
     * would add a class whose only job is to be unpacked one line later, and
     * would hide which vector fields this check actually reads.
     */
    @SuppressWarnings("java:S107")
    private static void checkSuccess(JsonNode scenario, int index, Map<String, Binding> bindings,
                                     JsonNode expect, String token, int generation, String familyId,
                                     long expiresAt, long idleExpiresAt) {
        if (expect == null) return;
        if (expect.has("generation") && expect.get("generation").asInt() != generation) {
            throw fail(scenario, index,
                    "expected generation " + expect.get("generation").asInt() + ", got " + generation);
        }
        if (expect.has("kid")) {
            String kid = token.split("\\.", -1)[1];
            if (!expect.get("kid").asText().equals(kid)) {
                throw fail(scenario, index, "expected kid " + expect.get("kid").asText() + ", got " + kid);
            }
        }
        if (expect.has("sameFamilyAs")) {
            Binding other = bindings.get(expect.get("sameFamilyAs").asText());
            if (other == null || !other.familyId().equals(familyId)) {
                throw fail(scenario, index, "familyId changed across rotation");
            }
        }
        if (expect.has("sameExpiresAtAs")) {
            Binding other = bindings.get(expect.get("sameExpiresAtAs").asText());
            if (other == null || other.expiresAt() != expiresAt) {
                throw fail(scenario, index, "absolute deadline moved: "
                        + (other == null ? "?" : other.expiresAt()) + " -> " + expiresAt);
            }
        }
        if (expect.path("idleEqualsExpires").asBoolean() && idleExpiresAt != expiresAt) {
            throw fail(scenario, index,
                    "idleExpiresAt " + idleExpiresAt + " should be clamped to " + expiresAt);
        }
    }

    /**
     * The tri-state [N-39] attribution expectation. {@code true} demands the
     * field, {@code false} demands its <em>absence</em> -- the exclusion list
     * (MALFORMED, UNKNOWN_KID, NOT_FOUND) is a requirement too, and the
     * truthy-only check this replaced could never observe it. A key the vector
     * omits asserts nothing.
     *
     * <p>Both failure records always declare the two components, so "absent"
     * reads as {@code null} here -- that is how the engine signals "no record
     * was resolved" on the refresh path and the revokeToken path alike.
     */
    private static void checkAttribution(JsonNode scenario, int index, JsonNode expect,
                                         String userId, String familyId) {
        if (expect == null) return;
        if (expect.has("hasUserId")) {
            boolean want = expect.get("hasUserId").asBoolean();
            if ((userId != null) != want) {
                throw fail(scenario, index,
                        "expected userId " + (want ? "present" : "absent") + " ([N-39])");
            }
        }
        if (expect.has("hasFamilyId")) {
            boolean want = expect.get("hasFamilyId").asBoolean();
            if ((familyId != null) != want) {
                throw fail(scenario, index,
                        "expected familyId " + (want ? "present" : "absent") + " ([N-39])");
            }
        }
    }

    private static void checkRevoked(JsonNode scenario, int index, JsonNode expect, int revoked) {
        if (expect != null && expect.has("revoked") && expect.get("revoked").asInt() != revoked) {
            throw fail(scenario, index,
                    "expected " + expect.get("revoked").asInt() + " revoked, got " + revoked);
        }
    }

    private static void bind(Map<String, Binding> bindings, JsonNode step,
                             String token, String familyId, long expiresAt) {
        if (step.has("bind")) {
            bindings.put(step.get("bind").asText(), new Binding(token, familyId, expiresAt));
        }
    }

    private static String resolveToken(JsonNode scenario, int index,
                                       Map<String, Binding> bindings, JsonNode ref) {
        if (ref == null) throw fail(scenario, index, "step has no token reference");
        if (ref.has("literal")) return ref.get("literal").asText();
        if (!ref.has("ref")) throw fail(scenario, index, "step has no token reference");

        Binding bound = bindings.get(ref.get("ref").asText());
        if (bound == null) throw fail(scenario, index, "unknown binding \"" + ref.get("ref").asText() + "\"");
        if (!ref.has("forge")) return bound.token();

        String[] parts = bound.token().split("\\.", -1);
        switch (ref.get("forge").asText()) {
            case "verifier" -> parts[3] = FORGED_VERIFIER;
            case "unknownKid" -> parts[1] = "zz";
            case "unknownSelector" -> parts[2] = FORGED_SELECTOR;
            default -> throw fail(scenario, index, "unknown forge \"" + ref.get("forge").asText() + "\"");
        }
        return String.join(".", parts);
    }

    /**
     * A JSON document cannot carry a lone surrogate, so the vectors name the
     * kind and the runner constructs it ([N-12]).
     */
    private static String deviceOf(JsonNode step) {
        if (step.has("deviceIdKind")) {
            String kind = step.get("deviceIdKind").asText();
            if ("lone-surrogate".equals(kind)) return LONE_SURROGATE;
            throw new IllegalArgumentException("unknown deviceIdKind \"" + kind + "\"");
        }
        return step.has("deviceId") ? step.get("deviceId").asText() : null;
    }

    private static String verifierPartOf(String token) {
        return token.split("\\.", -1)[3];
    }

    /**
     * Eight parameters (java:S107): the eight configuration settings a scenario
     * may override, one per parameter. They are exactly the {@code defaults}
     * keys of spec/behavior-vectors.json, so the signature is the vector schema.
     */
    @SuppressWarnings("java:S107")
    private static NebulaEngine build(JsonNode vectors, RefreshTokenStore store, AtomicLong now,
                                      long absoluteTtl, long idleTtl, long grace,
                                      List<String> kids, String activeKid) {
        Map<String, String> peppers = new LinkedHashMap<>();
        for (String kid : kids) peppers.put(kid, vectors.get("peppers").get(kid).asText());

        NebulaEngine.Config cfg = new NebulaEngine.Config();
        cfg.peppers = peppers;
        cfg.activeKid = activeKid;
        cfg.store = store;
        cfg.absoluteTtlSeconds = absoluteTtl;
        cfg.idleTtlSeconds = idleTtl;
        cfg.reuseGraceSeconds = grace;
        cfg.clock = now::get;
        return new NebulaEngine(cfg);
    }

    private static JsonNode setting(JsonNode defaults, JsonNode override, String name) {
        if (override != null && override.has(name)) return override.get(name);
        return defaults.get(name);
    }

    private static List<String> strings(JsonNode array) {
        List<String> out = new ArrayList<>();
        array.forEach(node -> out.add(node.asText()));
        return out;
    }

    private static AssertionError fail(JsonNode scenario, int index, String message) {
        List<String> requirements = strings(scenario.get("requirements"));
        return new AssertionError("[" + scenario.get("id").asText() + "] step " + index
                + " (" + String.join(", ", requirements) + "): " + message);
    }

    // --- store ---------------------------------------------------------------

    /**
     * Wraps the reference store so a scenario can force one compare-and-set to
     * lose, which is how the CONFLICT scenarios reproduce a concurrent refresh
     * deterministically ([N-17], [N-18]).
     */
    static final class ControllableStore implements RefreshTokenStore {

        final MemoryRefreshTokenStore inner = new MemoryRefreshTokenStore();
        private final Set<String> failNext = ConcurrentHashMap.newKeySet();

        void failNextCas(String method) {
            if (!Arrays.asList("markRotated", "revokeIfActive").contains(method)) {
                throw new IllegalArgumentException("not a compare-and-set: " + method);
            }
            failNext.add(method);
        }

        @Override
        public TokenRecord findBySelector(String selector) {
            return inner.findBySelector(selector);
        }

        @Override
        public void insert(TokenRecord row) {
            inner.insert(row);
        }

        @Override
        public boolean markRotated(String selector, TokenStatus fromStatus, long rotatedAt,
                                   String replacedBySelector) {
            if (failNext.remove("markRotated")) return false;
            return inner.markRotated(selector, fromStatus, rotatedAt, replacedBySelector);
        }

        @Override
        public boolean revokeIfActive(String selector) {
            if (failNext.remove("revokeIfActive")) return false;
            return inner.revokeIfActive(selector);
        }

        @Override
        public int revokeFamily(String familyId) {
            return inner.revokeFamily(familyId);
        }

        @Override
        public int revokeUser(String userId) {
            return inner.revokeUser(userId);
        }
    }
}
